### PR TITLE
Remove the --btf option

### LINKED ISSUE #1588: BTF and user types
Split from a thread in an old PR

----

So, after some experiments, I discovered a problem with this approach  - when asking BTF to dump a type definition, it also dumps definitions of all other types needed for successful compilation. If one of those types is overridden by the user, a re-definition error occurs (and it's quite difficult to discover which types were dumped).

This leads me to another thought - do we really want to allow user to do this? I'd say that allowing overriding a part of kernel types while taking the rest from BTF will always eventually blow up somewhere. I'd suggest enabling only the following scenarios:
- there are no user-defined types -> use BTF
- there are user-defined types but they do not override kernel types -> use BTF
- there are user-defined types conflicting with kernel types -> don't use BTF

This should allow user to define his own types that use existing kernel types without any problems, but if he wants to use types with conflicting names (e.g. for userspace tracing), he must define/include all the necessary types manually (or use them from `linux/types.h`, which is included in such case). This approach should be also backwards compatible for scripts that include kernel headers (such as the above `tcplife.bt`) as those must have already specified all the necessary types.

Any thoughs before I start implementing this?

_Originally posted by @viktormalik in https://github.com/iovisor/bpftrace/pull/1506#discussion_r516575033_

### PR DESCRIPTION
**Pull request by viktormalik**
This is an improved version of #1506, which caused type redefinition conflicts for some cases. Now, all the examples mentioned in the old PR should work.

BTF is parsed automatically if it is available and if the user didn't define any type (either via user definitions or via includes) that would redefine some type from BTF.

I think that this also closes #1588. It would seem nicer to allow the user to redefine a part of the types and take the rest from BTF, but I believe that could cause some unexpected behaviour (see [this comment](https://github.com/iovisor/bpftrace/issues/1588#issuecomment-721069928) for details).

The only problem of this implementation is that it checks for redefinitions by running the Clang parser and then checking the error messages. This is not a nice solution, but I haven't been able to find a better one, yet. If you have any ideas, please bring them in.

Also, I'm not sure if this solves the problem from #1511, as `linux/types.h` will be used when BTF is not and that may cause redefinition errors with `sys/types.h`. It shouldn't be hard to resolve this in a way similar to this PR, but I haven't been able to come with a reproducer.

##### Checklist

- [x] Language changes are updated in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by danobi** [comment]
Skimmed the code and it looks reasonable. I'll spend some time tomorrow to try and break it.

**Comment by danobi** [comment]
Can you write up the set of "rules" regarding how type definitions are chosen/acquired? Would help users understand errors and help us maintain behavior over time.

Also, there should be more runtime tests for this PR to test combinations of user-defined and BTF defined types. We can use ` REQUIRES_FEATURE btf ` and operate on `struct task_struct` (task_struct is unlikely to go away any time soon).

**Comment by danobi** [comment]
Seeing some errors with dcsnoop.bt:
```
sudo ./build/src/bpftrace tools/dcsnoop.bt
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/kasan-checks.h:18:9: error: use of undeclared identifier 'true'
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/kasan-checks.h:22:9: error: use of undeclared identifier 'true'
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/kasan-checks.h:36:9: error: use of undeclared identifier 'true'
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/kasan-checks.h:40:9: error: use of undeclared identifier 'true'
/lib/modules/5.9.14-100.fc32.x86_64/source/include/asm-generic/qspinlock_types.h:14:16: error: redefinition of 'qspinlock'
/bpftrace/include/__btf_generated_header.h:67:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/asm-generic/qspinlock_types.h:44:3: error: typedef redefinition with different types ('struct (anonymous struct at /lib/modules/5.9.14-100.fc32.x86_64/source/include/asm-generic/qspinlock_types.h:14:16)' vs 'struct qspinlock')
/bpftrace/include/__btf_generated_header.h:81:26: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/asm-generic/qrwlock_types.h:13:16: error: redefinition of 'qrwlock'
/bpftrace/include/__btf_generated_header.h:1415:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/asm-generic/qrwlock_types.h:27:3: error: typedef redefinition with different types ('struct (anonymous struct at /lib/modules/5.9.14-100.fc32.x86_64/source/include/asm-generic/qrwlock_types.h:13:16)' vs 'struct qrwlock')
/bpftrace/include/__btf_generated_header.h:1426:24: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/lockdep_types.h:187:8: error: redefinition of 'lock_class_key'
/bpftrace/include/__btf_generated_header.h:24:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/lockdep_types.h:192:8: error: redefinition of 'lockdep_map'
/bpftrace/include/__btf_generated_header.h:4566:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/spinlock_types.h:20:16: error: redefinition of 'raw_spinlock'
/bpftrace/include/__btf_generated_header.h:83:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/spinlock_types.h:29:3: error: typedef redefinition with different types ('struct (anonymous struct at /lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/spinlock_types.h:20:16)' vs 'struct raw_spinlock')
/bpftrace/include/__btf_generated_header.h:441:29: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/spinlock_types.h:71:16: error: redefinition of 'spinlock'
/bpftrace/include/__btf_generated_header.h:87:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/spinlock_types.h:83:3: error: typedef redefinition with different types ('struct (anonymous struct at /lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/spinlock_types.h:71:16)' vs 'struct spinlock')
/bpftrace/include/__btf_generated_header.h:537:25: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/rwlock_types.h:20:3: error: typedef redefinition with different types ('struct rwlock_t' vs 'struct rwlock_t')
/bpftrace/include/__btf_generated_header.h:1430:3: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/ratelimit_types.h:15:8: error: redefinition of 'ratelimit_state'
/bpftrace/include/__btf_generated_header.h:1622:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/jump_label.h:122:8: error: redefinition of 'jump_entry'
/bpftrace/include/__btf_generated_header.h:2382:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/time64.h:13:8: error: redefinition of 'timespec64'
/bpftrace/include/__btf_generated_header.h:2003:8: note: previous definition is here
/lib/modules/5.9.14-100.fc32.x86_64/source/include/linux/restart_block.h:16:6: error: redefinition of 'timespec_type'
/bpftrace/include/__btf_generated_header.h:397:6: note: previous definition is here
fatal error: too many errors emitted, stopping now [-ferror-limit=]
```

**Comment by viktormalik** [comment]
> Can you write up the set of "rules" regarding how type definitions are chosen/acquired? Would help users understand errors and help us maintain behavior over time.

Good idea. Actually, there are rules from the user point of view (i.e., what happens if I redefine a kernel type?) and from the implementation point of view (i.e., how the types are actually resolved?). I'd put the former into the reference guide and the latter as a comment into the code, sounds reasonable?

> Also, there should be more runtime tests for this PR to test combinations of user-defined and BTF defined types. We can use `REQUIRES_FEATURE btf` and operate on `struct task_struct` (task_struct is unlikely to go away any time soon).

Ok, I'll try to come with some tests.

**Comment by danobi** [comment]
> I'd put the former into the reference guide and the latter as a comment into the code, sounds reasonable?

Sounds great

**Comment by viktormalik** [comment]
Instructions for users concerning BTF types are [here](https://github.com/iovisor/bpftrace/pull/1669/files#diff-3891e9329d77f687dc1f8e6e025adbb10d664a439ba4e06fcfdd0b79df50db49R3287-R3305).
Details of type resolution implementation are [here](https://github.com/iovisor/bpftrace/pull/1669/files#diff-e922736b770aebf46d623db88058e892b9be72fde131604577d3c0007ef3df41R669-R695) (link may require to expand `clang_parser.cpp`).

I also added 2 runtime tests.

File: docs/reference_guide.md [hunk 1]
```diff
@@ -108,6 +108,7 @@ discussion to other files in /docs, the /tools/\*\_examples.txt files, or blog p
     - [1. `printf()`: Per-Event Output](#1-printf-per-event-output)
     - [2. `interval`: Interval Output](#2-interval-interval-output)
     - [3. `hist()`, `printf()`: Histogram Printing](#3-hist-print-histogram-printing)
+- [BTF Types](#btf-types)
```
Comment (danobi) [inline docs/reference_guide.md:111]: Maybe this is a bit more descriptive
```suggestion
- [BTF Support](#btf-types)
```

**Review by danobi — COMMENTED** [review]
LGTM, thanks for working on this. Small doc nit

File: docs/reference_guide.md (hunk 1 above)
Comment (viktormalik) [reply docs/reference_guide.md]: Right, I fixed it.

**Review by danobi — APPROVED** [review]

**Pull request merged**